packages/apps/plugins/plugin-markdown/src/components/EditorMain.tsx

```diff
@@ -2,48 +2,68 @@
 // Copyright 2023 DXOS.org
 //
 
-import React, { type HTMLAttributes, useMemo, useEffect } from 'react';
+import { type EditorView } from '@codemirror/view';
+import React, { useMemo, useEffect } from 'react';
 
 import { LayoutAction, useIntentResolver } from '@dxos/app-framework';
-import { useTranslation } from '@dxos/react-ui';
+import { useThemeContext, useTranslation } from '@dxos/react-ui';
 import {
-  type TextEditorProps,
   type Comment,
-  MarkdownEditor,
+  type TextEditorProps,
+  TextEditor,
   Toolbar,
-  cursorLineMargin,
+  createBasicExtensions,
+  createMarkdownExtensions,
+  createThemeExtensions,
+  decorateMarkdown,
   editorFillLayoutRoot,
   editorFillLayoutEditor,
   focusComment,
+  table,
+  image,
   useComments,
   useEditorView,
   useActionHandler,
   useFormattingState,
 } from '@dxos/react-ui-editor';
+import { formattingKeymap } from '@dxos/react-ui-editor';
 import { attentionSurface, focusRing, mx, textBlockWidth } from '@dxos/react-ui-theme';
+import { nonNullable } from '@dxos/util';
 
 import { MARKDOWN_PLUGIN } from '../meta';
 
+// Expose editor view for playwright tests.
+// TODO(wittjosiah): Find a better way to expose this or find a way to limit it to test runs.
+const useTest = (view: EditorView | null) => {
+  useEffect(() => {
+    const composer = (window as any).composer;
+    if (composer) {
+      composer.editorView = view;
+    }
+  }, [view]);
+};
+
 export type EditorMainProps = {
-  comments?: Comment[];
+  id: string;
+  readonly?: boolean;
   toolbar?: boolean;
-} & Pick<TextEditorProps, 'model' | 'readonly' | 'extensions'>;
+  comments?: Comment[];
+} & Pick<TextEditorProps, 'doc' | 'extensions'>;
 
-export const EditorMain = ({ model, comments, toolbar, extensions: _extensions, ...props }: EditorMainProps) => {
+// TODO(burdon): Invalid cursor error when switching between editors.
+export const EditorMain = ({ id, readonly, toolbar, comments, doc, extensions: _extensions }: EditorMainProps) => {
   const { t } = useTranslation(MARKDOWN_PLUGIN);
+  const { themeMode } = useThemeContext();
+
+  // TODO(burdon): Remove useEditorView. Check interaction with useComments with @wittjosiah.
```

Comment:
`viewInvalidated` is how we resolved the problem of attempting to render comment highlights from a new doc onto the previous doc. The ref doesn't change until after render has run but we can tie the ref to the id coming in such that we immediately know that a view is invalid and stop using it.